proving_system/stark/src/air/example/cairo.rs

```diff
@@ -0,0 +1,305 @@
+            - &curr[FRAME_OP0_ADDR];
+
+    constraints[OP1_ADDR] = &curr[F_OP_1_VAL] * pc
+        + &curr[F_OP_1_AP] * ap
+        + &curr[F_OP_1_FP] * fp
+        + (&one - &curr[F_OP_1_VAL] - &curr[F_OP_1_AP] - &curr[F_OP_1_FP]) * &curr[FRAME_OP0]
+        + (&curr[OFF_OP1] - &b15)
+        - &curr[FRAME_OP1_ADDR];
+}
+
+fn compute_register_constraints(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    let curr = frame.get_row(0);
+    let next = frame.get_row(1);
+
+    let one = FE::one();
+    let two = FE::from(2);
+
+    // ap and fp constraints
+    constraints[NEXT_AP] = &curr[FRAME_AP]
+        + &curr[F_AP_ADD] * &curr[FRAME_RES]
+        + &curr[F_AP_ONE]
+        + &curr[F_OPC_CALL] * &two
+        - &next[FRAME_AP];
+
+    constraints[NEXT_FP] = &curr[F_OPC_RET] * &curr[FRAME_DST]
+        + &curr[F_OPC_CALL] * (&curr[FRAME_AP] + &two)
+        + (&one - &curr[F_OPC_RET] - &curr[F_OPC_CALL]) * &curr[FRAME_FP]
+        - &next[FRAME_FP];
+
+    // pc constraints
+    constraints[NEXT_PC_1] = (&curr[FRAME_T1] - &curr[F_PC_JNZ])
+        * (&next[FRAME_PC] - (&curr[FRAME_PC] + frame_inst_size(curr)));
+
+    constraints[NEXT_PC_2] = &curr[FRAME_T0]
+        * (&next[FRAME_PC] - (&curr[FRAME_PC] + &curr[FRAME_OP1]))
+        + (&one - &curr[F_PC_JNZ]) * &next[FRAME_PC]
+        - ((&one - &curr[F_PC_ABS] - &curr[F_PC_REL] - &curr[F_PC_JNZ])
+            * (&curr[FRAME_PC] + frame_inst_size(curr))
+            + &curr[F_PC_ABS] * &curr[FRAME_RES]
+            + &curr[F_PC_REL] * (&curr[FRAME_PC] + &curr[FRAME_RES]));
+
+    constraints[T0] = &curr[F_PC_JNZ] * &curr[FRAME_DST] - &curr[FRAME_T0];
+    constraints[T1] = &curr[FRAME_T0] * &curr[FRAME_RES] - &curr[FRAME_T1];
+}
+
+fn compute_opcode_constraints(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    let curr = frame.get_row(0);
+    let one = FE::one();
+
+    constraints[MUL_1] = &curr[FRAME_MUL] - (&curr[FRAME_OP0] * &curr[FRAME_OP1]);
+
+    constraints[MUL_2] = &curr[F_RES_ADD] * (&curr[FRAME_OP0] + &curr[FRAME_OP1])
+        + &curr[F_RES_MUL] * &curr[FRAME_MUL]
+        + (&one - &curr[F_RES_ADD] - &curr[F_RES_MUL] - &curr[F_PC_JNZ]) * &curr[FRAME_OP1]
+        - (&one - &curr[F_PC_JNZ]) * &curr[FRAME_RES];
+
+    constraints[CALL_1] = &curr[F_OPC_CALL] * (&curr[FRAME_DST] - &curr[FRAME_FP]);
+
+    constraints[CALL_2] =
+        &curr[F_OPC_CALL] * (&curr[FRAME_OP0] - (&curr[FRAME_PC] + frame_inst_size(curr)));
+
+    constraints[ASSERT_EQ] = &curr[F_OPC_AEQ] * (&curr[FRAME_DST] - &curr[FRAME_RES]);
+}
+
+fn enforce_selector(constraints: &mut [FE], frame: &Frame<Stark252PrimeField>) {
+    let curr = frame.get_row(0);
+    for result_cell in constraints.iter_mut().take(ASSERT_EQ + 1).skip(INST) {
+        *result_cell = result_cell.clone() * curr[FRAME_SELECTOR].clone();
+    }
+}
+
+fn frame_inst_size(frame_row: &[FE]) -> FE {
+    &frame_row[F_OP_1_VAL] + FE::one()
+}
```

Comment:
We could add here a test that imports the trace of a simple program and checks that all constraints are zero.

Reply:
done, thanks!